File: lib_common/src/main/java/com/ebook/common/view/ReadBookControl.java

package com.ebook.common.view;

import android.graphics.Color;

import com.blankj.utilcode.util.SPUtils;
import com.ebook.common.R;
import com.ebook.common.util.DisplayUtil;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ReadBookControl {
    public static final int DEFAULT_TEXT = 2;
    public static final int DEFAULT_BG = 1;

    private static List<Map<String,Integer>> textKind;
    private static List<Map<String,Integer>> textDrawable;

    private int textSize;
    private int textExtra;
    private int textColor;
    private int textBackground;

    private int textKindIndex = DEFAULT_TEXT;
    private int textDrawableIndex = DEFAULT_BG;

    private Boolean canClickTurn = true;
    private Boolean canKeyTurn = true;

    private SPUtils preference;

    private static ReadBookControl readBookControl;

    public static ReadBookControl getInstance(){
        if(readBookControl == null){
            synchronized (ReadBookControl.class){
                if(readBookControl == null){
                    readBookControl = new ReadBookControl();
                }
            }
        }
        return readBookControl;
    }
    private ReadBookControl(){
        if(null == textKind){
            textKind = new ArrayList<>();
            Map<String,Integer> temp1 = new HashMap<>();
            temp1.put("textSize", 14);
            temp1.put("textExtra", DisplayUtil.dip2px(6.5f));
            textKind.add(temp1);

            Map<String,Integer> temp2 = new HashMap<>();
            temp2.put("textSize", 16);
            temp2.put("textExtra", DisplayUtil.dip2px(8));
            textKind.add(temp2);

            Map<String,Integer> temp3 = new HashMap<>();
            temp3.put("textSize", 17);
            temp3.put("textExtra", DisplayUtil.dip2px(9));
            textKind.add(temp3);

            Map<String,Integer> temp4 = new HashMap<>();
            temp4.put("textSize", 20);
            temp4.put("textExtra", DisplayUtil.dip2px(11));
            textKind.add(temp4);

            Map<String,Integer> temp5 = new HashMap<>();
            temp5.put("textSize", 22);
            temp5.put("textExtra", DisplayUtil.dip2px(13));
            textKind.add(temp5);
        }
        if(null == textDrawable){
            textDrawable = new ArrayList<>();
            Map<String,Integer> temp1 = new HashMap<>();
            temp1.put("textColor",Color.parseColor("#3E3D3B"));
            temp1.put("textBackground", R.drawable.shape_bg_readbook_white);
            textDrawable.add(temp1);

            Map<String,Integer> temp2 = new HashMap<>();
            temp2.put("textColor",Color.parseColor("#5E432E"));
            temp2.put("textBackground", R.drawable.bg_readbook_yellow);
            textDrawable.add(temp2);

            Map<String,Integer> temp3 = new HashMap<>();
            temp3.put("textColor",Color.parseColor("#22482C"));
            temp3.put("textBackground",R.drawable.bg_readbook_green);
            textDrawable.add(temp3);

            Map<String,Integer> temp4 = new HashMap<>();
            temp4.put("textColor",Color.parseColor("#808080"));
            temp4.put("textBackground",R.drawable.bg_readbook_black);
            textDrawable.add(temp4);
        }
        preference = SPUtils.getInstance("CONFIG");
        this.textKindIndex = preference.getInt("textKindIndex",DEFAULT_TEXT);
        this.textSize = textKind.get(textKindIndex).get("textSize");
        this.textExtra = textKind.get(textKindIndex).get("textExtra");
        this.textDrawableIndex = preference.getInt("textDrawableIndex",DEFAULT_BG);
        this.textColor = textDrawable.get(textDrawableIndex).get("textColor");
        this.textBackground = textDrawable.get(textDrawableIndex).get("textBackground");

        this.canClickTurn = preference.getBoolean("canClickTurn",true);
        this.canKeyTurn = preference.getBoolean("canClickTurn",true);
    }

    public int getTextSize() {
        return textSize;
    }

    public int getTextExtra() {
        return textExtra;
    }

    public int getTextColor() {
        return textColor;
    }

    public int getTextBackground() {
        return textBackground;
    }

    public int getTextKindIndex() {
        return textKindIndex;
    }

    public void setTextKindIndex(int textKindIndex) {
        this.textKindIndex = textKindIndex;
        preference.put("textKindIndex",textKindIndex);
        this.textSize = textKind.get(textKindIndex).get("textSize");
        this.textExtra = textKind.get(textKindIndex).get("textExtra");
    }

    public int getTextDrawableIndex() {
        return textDrawableIndex;
    }

    public void setTextDrawableIndex(int textDrawableIndex) {
        this.textDrawableIndex = textDrawableIndex;
        preference.put("textDrawableIndex",textDrawableIndex);
        this.textColor = textDrawable.get(textDrawableIndex).get("textColor");
        this.textBackground = textDrawable.get(textDrawableIndex).get("textBackground");
    }

    public static List<Map<String, Integer>> getTextKind() {
        return textKind;
    }

    public static List<Map<String, Integer>> getTextDrawable() {
        return textDrawable;
    }

    public Boolean getCanKeyTurn() {
        return canKeyTurn;
    }

    public void setCanKeyTurn(Boolean canKeyTurn) {
        this.canKeyTurn = canKeyTurn;
       preference.put("canKeyTurn",canKeyTurn);
    }

    public Boolean getCanClickTurn() {
        return canClickTurn;
    }

    public void setCanClickTurn(Boolean canClickTurn) {
        this.canClickTurn = canClickTurn;
        preference.put("canClickTurn",canClickTurn);
    }
}

File: lib_book_common/src/main/java/com/ebook/common/view/ReadBookControl.kt
package com.ebook.common.view

import androidx.core.graphics.toColorInt
import com.ebook.common.R
import com.ebook.common.util.SPUtil
import com.xrn1997.common.util.DisplayUtil

/**
 * ReadBookControl
 * 内存缓存 + SP 持久化，避免频繁 I/O。
 */
object ReadBookControl {

    private const val SP_NAME = "CONFIG"
    const val DEFAULT_TEXT = 2
    const val DEFAULT_BG = 1

    /** 字体样式 */
    data class TextKind(val textSize: Int, val textExtra: Int)

    /** 背景主题 */
    data class TextDrawable(val textColor: Int, val textBackground: Int)

    /** 字体样式列表（初始化一次） */
    private val textKindList: List<TextKind> = listOf(
        TextKind(14, DisplayUtil.dip2px(6.5f).toInt()),
        TextKind(16, DisplayUtil.dip2px(8f).toInt()),
        TextKind(17, DisplayUtil.dip2px(9f).toInt()),
        TextKind(20, DisplayUtil.dip2px(11f).toInt()),
        TextKind(22, DisplayUtil.dip2px(13f).toInt()),
        TextKind(24, DisplayUtil.dip2px(15f).toInt()),
        TextKind(26, DisplayUtil.dip2px(17f).toInt()),
        TextKind(30, DisplayUtil.dip2px(21f).toInt())
    )

    /** 背景主题列表（初始化一次） */
    private val textDrawableList: List<TextDrawable> = listOf(
        TextDrawable("#3E3D3B".toColorInt(), "#F3F3F3".toColorInt()),//白
        TextDrawable("#5E432E".toColorInt(), "#DCD1BC".toColorInt()),//黄
        TextDrawable("#22482C".toColorInt(), "#E1F0D9".toColorInt()),//绿
        TextDrawable("#808080".toColorInt(), "#2D2D33".toColorInt())//黑
    )

    // ----------------------------
    // 内存缓存属性
    // ----------------------------
    var textKindIndex: Int
        private set
    var textDrawableIndex: Int
        private set
    var textSize: Int
        private set
    var textExtra: Int
        private set
    var textColor: Int
        private set
    var textBackground: Int
        private set
    var canClickTurn: Boolean
        private set
    var canKeyTurn: Boolean
        private set

    init {
        // 从 SP 初始化（只读一次）
        textKindIndex = SPUtil.get("textKindIndex", DEFAULT_TEXT, SP_NAME)
        if (textKindIndex !in textKindList.indices) textKindIndex = DEFAULT_TEXT
        textSize = textKindList[textKindIndex].textSize
        textExtra = textKindList[textKindIndex].textExtra

        textDrawableIndex = SPUtil.get("textDrawableIndex", DEFAULT_BG, SP_NAME)
        if (textDrawableIndex !in textDrawableList.indices) textDrawableIndex = DEFAULT_BG
        textColor = textDrawableList[textDrawableIndex].textColor
        textBackground = textDrawableList[textDrawableIndex].textBackground

        canClickTurn = SPUtil.get("canClickTurn", true, SP_NAME)
        canKeyTurn = SPUtil.get("canKeyTurn", true, SP_NAME)
    }

    // ----------------------------
    // 更新方法（避免 setter 与属性名冲突）
    // ----------------------------

    fun updateTextKindIndex(index: Int) {
        if (index !in textKindList.indices) return
        textKindIndex = index
        textSize = textKindList[index].textSize
        textExtra = textKindList[index].textExtra
        SPUtil.put("textKindIndex", index, SP_NAME)
    }

    fun updateTextDrawableIndex(index: Int) {
        if (index !in textDrawableList.indices) return
        textDrawableIndex = index
        textColor = textDrawableList[index].textColor
        textBackground = textDrawableList[index].textBackground
        SPUtil.put("textDrawableIndex", index, SP_NAME)
    }

    fun setCanClickTurn(enable: Boolean) {
        canClickTurn = enable
        SPUtil.put("canClickTurn", enable, SP_NAME)
    }

    fun setCanKeyTurn(enable: Boolean) {
        canKeyTurn = enable
        SPUtil.put("canKeyTurn", enable, SP_NAME)
    }

    // ----------------------------
    // 工具方法
    // ----------------------------
    fun getCurrentTextKind(): TextKind = textKindList[textKindIndex]
    fun getCurrentTextDrawable(): TextDrawable = textDrawableList[textDrawableIndex]
    fun getTextKindList(): List<TextKind> = textKindList
    fun getTextDrawableList(): List<TextDrawable> = textDrawableList
}
